cpg_workflows/stages/realign_genotype_with_dragen.py

```diff
@@ -0,0 +1,357 @@
+
+from hailtop.batch.job import PythonJob
+
+import cpg_utils
+from cpg_utils.cloud import get_path_components_from_gcp_path
+from cpg_utils.config import config_retrieve, image_path
+from cpg_utils.hail_batch import Batch, authenticate_cloud_credentials_in_job, get_batch
+from cpg_workflows.stages.dragen_ica import (
+    monitor_align_genotype_with_dragen,
+    prepare_ica_for_analysis,
+    run_align_genotype_with_dragen,
+    upload_data_to_ica,
+)
+from cpg_workflows.targets import SequencingGroup
+from cpg_workflows.workflow import SequencingGroupStage, StageInput, StageOutput, stage
+
+if TYPE_CHECKING:
+    from hailtop.batch.job import BashJob, PythonJob
+
+DRAGEN_VERSION: Final = config_retrieve(['ica', 'pipelines', 'dragen_version'])
+GCP_FOLDER_FOR_ICA_PREP: Final = f'ica/{DRAGEN_VERSION}/prepare'
+GCP_FOLDER_FOR_RUNNING_PIPELINE: Final = f'ica/{DRAGEN_VERSION}/pipelines'
+GCP_FOLDER_FOR_ICA_DOWNLOAD: Final = f'ica/{DRAGEN_VERSION}/output'
+ICA_REST_ENDPOINT: Final = 'https://ica.illumina.com/ica/rest'
+
+
+coloredlogs.install(level=logging.INFO)
+
+
+def calculate_needed_storage(
+    cram: str,
+    bucket_name: str,
+) -> str:
+    logging.info(f'Checking blob size for {cram}')
+    storage_client = storage.Client()
+    gcp_bucket = storage_client.bucket(bucket_name=bucket_name)
+    blob_to_upload_size_bytes: int = gcp_bucket.get_blob(blob_name=f'{cram}').size
+    storage_size: int = ceil(ceil((blob_to_upload_size_bytes / (1024**3)) + 3) * 1.2)
+    logging.info(f'Calculated storage is {storage_size}Gi.')
+    return f'{storage_size}Gi'
+
+
+# No need to register this stage in Metamist I think, just ICA prep
+@stage(analysis_keys=['cram_fid', 'cram_index_fid', 'analysis_output_fid'])
+class PrepareIcaForDragenAnalysis(SequencingGroupStage):
+    """Set up ICA for a single realignment run.
+
+    Creates a file ID for both the CRAM and CRAI file to upload to.
+    Creates a folder ID for the Dragen output to be written into.
+
+    """
+
+    def expected_outputs(self, sequencing_group: SequencingGroup) -> dict[str, cpg_utils.Path]:
+        sg_bucket: str = f'{get_path_components_from_gcp_path(str(sequencing_group.cram))["bucket"]}'
+        output_dict = {
+            'cram_fid': cpg_utils.to_path(
+                f'gs://{sg_bucket}/{GCP_FOLDER_FOR_ICA_PREP}/{sequencing_group.name}.cram_ica_file_id',
+            ),
+            'crai_fid': cpg_utils.to_path(
+                f'gs://{sg_bucket}/{GCP_FOLDER_FOR_ICA_PREP}/{sequencing_group.name}.cram.crai_ica_file_id',
+            ),
+            'analysis_output_fid': cpg_utils.to_path(
+                f'gs://{sg_bucket}/{GCP_FOLDER_FOR_ICA_PREP}/{sequencing_group.name}_dragen_output_folder_id',
+            ),
+        }
+        return output_dict
+
+    def queue_jobs(self, sequencing_group: SequencingGroup, inputs: StageInput) -> StageOutput | None:
+        cram_path_components: dict[str, str] = get_path_components_from_gcp_path(path=str(object=sequencing_group.cram))
```

Comment:
There's a lot of use of `get_path_components_from_gcp_path`, the method is fine, but it feels like a long way round.

Instead of breaking into component parts and getting the "file" part, you could just use `sequencing_group.cram.name` for instance, which is much easier to read.

In general the Path objects have all the pieces you're looking for as instance attributes

```
cram_path_components['file'] == sequencing_group.cram.name
cram_path_components['bucket'] == sequencing_group.cram.drive
{cram_path_components["suffix"]}{cram_path_components["file"]} == sequencing_group.cram.blob
```

Reply:
Cloudpathlib (the result of `to_path('gs://something')`, returned by most of our helper methods, and PathLib have the same interface, so this would be a decent doc for all the various path elements these Path objects will have available as properties:

https://docs.python.org/3/library/pathlib.html#methods-and-properties

Reply:
I believe I've resolved all of these, except where strictly necessary